run-spill-build/overlaySinglesIntoSpillsSortedWithNuIntTime.cpp

```diff
@@ -0,0 +1,513 @@
+    // instead we do: first time -> corresponding event 
+    for (const auto& ttime : times) {
+      bool is_nu = ttime.tag == 1;
+
+      TTree* in_tree = is_nu ? edep_evts_1 : edep_evts_2;
+      TTree* gn_tree = is_nu ? genie_evts_1 : genie_evts_2;
+
+      int& evt_it = is_nu ? evt_it_1 : evt_it_2;
+
+      in_tree->GetEntry(ttime.evId);
+      gn_tree->GetEntry(ttime.evId);
+
+      TG4Event* edep_evt = is_nu ? edep_evt_1 : edep_evt_2;
+      // out_branch->ResetAddress();
+      out_branch->SetAddress(&edep_evt); // why the &? what is meaning of life
+      // new_tree->SetBranchAddress("Event", &edep_evt);
+
+      int globalSpillId = int(1E3)*spillFileId + spillN;
+
+      std::string event_string = std::to_string(edep_evt->RunId) + " "
+        + std::to_string(edep_evt->EventId);
+      std::string spill_string = std::to_string(globalSpillId);
+      TObjString* event_tobj = new TObjString(event_string.c_str());
+      TObjString* spill_tobj = new TObjString(spill_string.c_str());
+
+      if (event_spill_map->FindObject(event_string.c_str()) == 0)
+        event_spill_map->Add(event_tobj, spill_tobj);
+      else {
+        std::cerr << "[ERROR] redundant event ID " << event_string.c_str() << std::endl;
+        std::cerr << "event_spill_map entries = " << event_spill_map->GetEntries() << std::endl;
+        throw;
+      }
+
+      double event_time = ttime.time + conversionTo_ns*spillPeriod_s*spillN;
+      double old_event_time = 0.;
+
+      gRooTracker& genie_evts_data = is_nu ? genie_evts_1_data : genie_evts_2_data;
+      genie_tree_data.CopyFrom(genie_evts_data);
+      genie_tree_data.EvtNum = edep_evt->EventId;
+      genie_tree_data.EvtVtx[3] = event_time;
+
+      int nPrimaryPartThisEvent = 0;
+      for (auto &v : edep_evt->Primaries)
+        nPrimaryPartThisEvent += v.Particles.size();
+      int nTrajectoriesThisEvent = edep_evt->Trajectories.size();
+      int nSecondaryPartThisEvent = nTrajectoriesThisEvent - nPrimaryPartThisEvent;
+
+      auto updateTrackId = [lastPriTrajId, lastSecTrajId, nPrimaryPartThisEvent](int &trkId)
+      { trkId = trkId < nPrimaryPartThisEvent ? lastPriTrajId + trkId : lastSecTrajId + trkId - nPrimaryPartThisEvent; };
+
+      // ... interaction vertex
+      for (std::vector<TG4PrimaryVertex>::iterator v = edep_evt->Primaries.begin(); v != edep_evt->Primaries.end(); ++v) {
+        old_event_time = v->Position.T();
+        v->Position.SetT(event_time);
+        // TMS reco wants the InteractionNumber to be the entry number of the
+        // vertex in DetSimPassThru/gRooTracker. Since, by construction, our
+        // EDepSimEvents and gRooTracker trees are aligned one-to-one, we just
+        // trivially set InteractionNumber to be the current entry number.
+        // https://github.com/DUNE/2x2_sim/issues/54
+        v->InteractionNumber = evt_it_1 + evt_it_2;
+        for (auto &p : v->Particles){
+          updateTrackId(p.TrackId);
+        }
+      }
```

Comment:
If there is the assumption that Primaries contains just one element, why we need to loop over it?